### ISSUE TITLE
Cannot build our project with cabal-3.0, ghc-8.8.1, and cassava

### ISSUE THREAD

**Issue by recursion-ninja** [haskell-hvr/cassava#177]
`cassava` has been causing compilation problems for our project for a long time since transitioning to `ghc-8.8.1` in July. I excised our direct usage of the library but this did not resolve the compilation issues, as `cassava` is a transitive dependency of many packages we depend on and cannot remove.

You can replicate the results as follows:

```
$ cabal --version
cabal-install version 3.0.0.0
compiled using version 3.0.0.0 of the Cabal library
$ git clone https://github.com/amnh/PCG
$ cd PCG
$ git checkout a7591a14b8747a7137bcfa
$ cabal new-build --with-compiler=ghc-8.8.1 --allow-newer
```

Which results in the following output:

```
Cloning into '/home/user/dir/pcg-clean/PCG/dist-newstyle/src/vector-th_-ba3e92aaa74b47d7'...
remote: Enumerating objects: 6, done.
remote: Counting objects: 100% (6/6), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 265 (delta 0), reused 4 (delta 0), pack-reused 259
Receiving objects: 100% (265/265), 37.59 KiB | 2.35 MiB/s, done.
Resolving deltas: 100% (97/97), done.
HEAD is now at 2fa686c support template-haskell-2.15
Resolving dependencies...
Build profile: -w ghc-8.8.1 -O1
In order, the following will be built (use -v for more details):
 - cassava-0.5.2.0 (lib) (requires build)
 - pcg-exportable-0.2.0 (lib) (first run)
 - vector-th-unbox-0.2.1.6 (lib) (first run)
 - math-functions-0.3.3.0 (lib) (requires build)
 - mwc-random-0.14.0.0 (lib) (requires build)
 - dense-linear-algebra-0.1.0.0 (lib) (requires build)
 - monad-par-0.3.4.8 (lib) (requires build)
 - foldl-1.4.5 (lib) (requires build)
 - clustering-0.4.1 (lib) (requires build)
 - statistics-0.15.1.1 (lib) (requires build)
 - mmark-0.0.7.1 (lib) (requires build)
 - criterion-1.5.6.1 (lib) (requires build)
 - pcg-validation-transformer-0.2.0 (lib) (first run)
 - pcg-utility-0.2.0 (lib) (first run)
 - pcg-tcm-memo-0.2.0 (lib) (first run)
 - pcg-serialize-0.2.0 (lib) (first run)
 - pcg-evaluation-0.2.0 (lib) (first run)
 - phylogenetic-component-graph-0.2.0 (exe:graphviz-examples) (first run)
 - phylogenetic-component-graph-0.2.0 (exe:generate-tcm) (first run)
 - pcg-file-source-0.2.0 (lib) (first run)
 - pcg-core-0.2.0 (lib:pcg-tcm) (first run)
 - pcg-alphabet-0.2.0 (lib) (first run)
 - pcg-language-0.2.0 (lib) (first run)
 - phylogenetic-component-graph-0.2.0 (exe:generate-data-set) (first run)
 - pcg-file-parsers-0.2.0 (lib) (first run)
 - pcg-core-0.2.0 (lib:pcg-data-structures) (first run)
 - phylogenetic-component-graph-0.2.0 (exe:newick-resolver) (first run)
 - phylogenetic-component-graph-0.2.0 (exe:fasta-differ) (first run)
 - pcg-core-0.2.0 (lib:pcg-analysis) (first run)
 - pcg-core-0.2.0 (lib) (first run)
 - phylogenetic-component-graph-0.2.0 (exe:ukkonen-do-comparison) (first run)
 - phylogenetic-component-graph-0.2.0 (exe:affine-safety) (first run)
 - phylogenetic-component-graph-0.2.0 (exe:3d-do-comparison) (first run)
 - phylogenetic-component-graph-0.2.0 (exe:2d-do-comparison) (first run)
 - pcg-data-normalization-0.2.0 (lib) (first run)
 - pcg-data-unification-0.2.0 (lib) (first run)
 - phylogenetic-component-graph-0.2.0 (exe:pcg) (first run)
Starting     cassava-0.5.2.0 (lib)
Configuring library for pcg-exportable-0.2.0..
Configuring library for vector-th-unbox-0.2.1.6..
Preprocessing library for pcg-exportable-0.2.0..
Building library for pcg-exportable-0.2.0..
Building     cassava-0.5.2.0 (lib)
[1 of 2] Compiling Bio.Character.Exportable.Class
Preprocessing library for vector-th-unbox-0.2.1.6..
Building library for vector-th-unbox-0.2.1.6..
[1 of 1] Compiling Data.Vector.Unboxed.Deriving ( Data/Vector/Unboxed/Deriving.hs, /home/user/dir/pcg-clean/PCG/dist-newstyle/build/x86_64-linux/ghc-8.8.1/vector-th-unbox-0.2.1.6/build/Data/Vector/Unboxed/Deriving.o )
[2 of 2] Compiling Bio.Character.Exportable

Failed to build cassava-0.5.2.0.
Build log (
/home/user/.cabal/logs/ghc-8.8.1/cassava-0.5.2.0-161ff8506acb50707cb4cacaf23ba2e0cf91ae301f7c34cca2314dc0ee5fc2d3.log
):
Configuring library for cassava-0.5.2.0..
Preprocessing library for cassava-0.5.2.0..
Building library for cassava-0.5.2.0..

src/Data/Csv/Conversion.hs:26:3: error:
     error: #error **INVARIANT BROKEN** Detected invalid combination of `text-short` and `bytestring` versions. Please verify the `pre-bytestring-0.10-4` flag-logic in the .cabal file wasn't elided.
     # error **INVARIANT BROKEN** Detected invalid combination of `text-short` and `bytestring` versions. Please verify the `pre-bytestring-0.10-4` flag-logic in the .cabal file wasn't elided.
       ^~~~~
   |
26 | # error **INVARIANT BROKEN** Detected invalid combination of `text-short` and `bytestring` versions. Please verify the `pre-bytestring-0.10-4` flag-logic in the .cabal file wasn't elided.
   |   ^

src/Data/Csv/Conversion.hs:100:0: error:
     warning: "MIN_VERSION_text_short" is not defined, evaluates to 0 [-Wundef]
     #if MIN_VERSION_text_short(0,1,0)
     
    |
100 | #if MIN_VERSION_text_short(0,1,0)
    | ^

src/Data/Csv/Conversion.hs:100:0: error:
     error: missing binary operator before token "("
    |
100 | #if MIN_VERSION_text_short(0,1,0)
    | ^

src/Data/Csv/Conversion.hs:1031:0: error:
     warning: "MIN_VERSION_text_short" is not defined, evaluates to 0 [-Wundef]
     #if MIN_VERSION_text_short(0,1,0)
     
     |
1031 | #if MIN_VERSION_text_short(0,1,0)
     | ^

src/Data/Csv/Conversion.hs:1031:0: error:
     error: missing binary operator before token "("
     |
1031 | #if MIN_VERSION_text_short(0,1,0)
     | ^

src/Data/Csv/Conversion.hs:1:1: error:
    `gcc' failed in phase `C pre-processor'. (Exit code: 1)
  |
1 | {-# LANGUAGE
  | ^
cabal: Failed to build cassava-0.5.2.0 (which is required by
exe:ukkonen-do-comparison from phylogenetic-component-graph-0.2.0, exe:pcg
from phylogenetic-component-graph-0.2.0 and others). See the build log above
for details.
```

As far as I can tell, `cassava` is the root of the problem. Can you help me resolve the issue?

**Comment by RyanGlScott** [comment]
I'm not the maintainer of `cassava`, but I have encountered this issue several times before. As a result, I'm pretty accustomed to working around it by now. Here is my (unofficial) diagnosis and recommended workaround.

The immediate reason why you're encountering this error is because you're passing `--allow-newer`, which essentially causes all upper version bounds on all build dependencies to be ignored. One consequence of this is that the upper version bounds on `bytestring` are ignored, which violates an assumption that `cassava` makes use of to ensure that only valid combinations of `bytestring` and related libraries are used. (I'm not super-familiar with how this assumption works, so this is a crude explanation at best.)

In other words, `cassava` simply cannot build with `--allow-newer=bytestring`. Nor can it build with `--allow-newer`, since that flag implies `--allow-newer=bytestring` (on top of `--allow-newer=<every-other-library>`).

For this reason, I discourage the use of `--allow-newer` and encourage only ignoring version bounds on individual packages. For instance, I was able to build your `PCG` repo with the following command:

```
$ cabal build --with-compiler=ghc-8.8.1 --allow-newer=base,polyparse
```

I had to ignore upper version bounds on `base` due to ocharles/tasty-rerun#14 and on `polyparse` due to ivan-m/graphviz#47, but everything else seemed to be fine.

**Comment by recursion-ninja** [comment]
@RyanGlScott thank you for the explaination. It has been **very** helpful. The build configuration of `cassava` is so complex, and the error messages reported are so useless, I had no idea why the package continued to build with such fragility. 

I must say that this fragility is quite burdensome when trying bleeding edge compilers and *many*, **many** upper version bounds must be ignored due to the common convention of having tight upper version bounds on `base`. The requirement to test for and then specify `allow-newer` explicitly for all dependencies which require it (of which we have 175) is a usability nightmare introduced only because `cassava` has fragile builds with the general `--allow-newer` flag. 

I seriously question if the CPP, cabal flags, etc. are really worth the trouble and fragility of builds, just for `cassava` to support extensive backwards comparability with what I would consider to be antiquated or obsoleted versions of `ghc`, `bytestring`, and `text-short`.

I know that you @RyanGlScott , are not the maintainer, but I would like to ask @hvr to consider releasing a `cassava-6.0.0` that drops the extensive backwards compatibility in favor of a much simpler and more robust build configuration. If anyone really needs the backwards compatibility, they can use `cassava < 6.0.0` as a build constraint. Would this be a reasonable path forward in the near future?

**Comment by GeorgeCo** [comment]
fwiw, this issue still exists with ghc version  8.10.0.20191210. Hopefully this will be resolved with newer versions of bytestring and/or ghc.

**Comment by hvr** [comment]
> In other words, `cassava` simply cannot build with `--allow-newer=bytestring`.

That's a general caveat of `--allow-newer`; when you use `--allow-newer`, you're also responsible for making sure the cabal flags whose constraints this might undermine are set properly.


>  that drops the extensive backwards compatibility in favor of a much simpler and more robust build configuration

The build configuration *is* robust; but `--allow-newer` tends to disrupt carefully crafted package configurations relying on cabal flags for selecting configurations; it's a bit like `unsafePerformIO` which undermines the typechecker and transfers more responsibility to the programmer.

Here's a more principled use of `--allow-newer` which is more surgical rather than the naive "ignore all constraints on bytestring regardless of any invariants" sledge-hammer:

```
--allow-newer="cassava-0.5.2.0:bytestring"
--constraint="cassava -bytestring--lt-0_10_4"
--constraint="bytestring >= 0.10.4"
```

**Comment by andreasabel** [comment]
Question seems answered.  Closing.  Reopen with new information if necessary.
